temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+                    details = await self._client.data_converter.decode(
+                        cancel_attr.details.payloads
+                    )
+                raise WorkflowFailureError(
+                    cause=temporalio.failure.FailureError(
+                        "Workflow cancelled",
+                        temporalio.failure.CancelledFailure(*details),
+                    )
+                )
+            elif event.HasField("workflow_execution_terminated_event_attributes"):
+                term_attr = event.workflow_execution_terminated_event_attributes
+                details = []
+                if term_attr.details and term_attr.details.payloads:
+                    details = await self._client.data_converter.decode(
+                        term_attr.details.payloads
+                    )
+                raise WorkflowFailureError(
+                    cause=temporalio.failure.FailureError(
+                        term_attr.reason or "Workflow terminated",
+                        temporalio.failure.TerminatedFailure(
+                            *details,
+                            reason=term_attr.reason or None,
+                        ),
+                    )
+                )
+            elif event.HasField("workflow_execution_timed_out_event_attributes"):
+                time_attr = event.workflow_execution_timed_out_event_attributes
+                # Follow execution
+                if follow_runs and time_attr.new_execution_run_id:
+                    req.execution.run_id = time_attr.new_execution_run_id
+                    req.next_page_token = b""
+                    continue
+                raise WorkflowFailureError(
+                    cause=temporalio.failure.FailureError(
+                        "Workflow timed out",
+                        temporalio.failure.TimeoutFailure(
+                            temporalio.failure.TimeoutType.START_TO_CLOSE
+                        ),
+                    )
+                )
+            elif event.HasField("workflow_execution_continued_as_new_event_attributes"):
+                cont_attr = event.workflow_execution_continued_as_new_event_attributes
+                if not cont_attr.new_execution_run_id:
+                    raise RuntimeError(
+                        "Unexpectedly missing new run ID from continue as new"
+                    )
+                # Follow execution
+                if follow_runs:
+                    req.execution.run_id = cont_attr.new_execution_run_id
+                    req.next_page_token = b""
+                    continue
+                raise WorkflowContinuedAsNewError(cont_attr.new_execution_run_id)
+            else:
+                raise RuntimeError(
+                    f"Unexpected close event attribute of {event.WhichOneof('attributes')}"
+                )
+
+    async def cancel(
+        self,
+        *,
+        run_id: Optional[str] = SELF_RUN_ID,
```

Comment:
Same here and the methods below, the handle already has a run_id.

Reply:
The handle _may_ already have a run ID. It may not. I allow you to set parameters of the call if you so choose on a per method level instead of having to go get another handle. What is the harm of allowing you to set call parameters in the call?

Reply:
> I'd rather have 1 way to do something than 2.

This absolutist position is unfair to apply everywhere. Otherwise, we'd sure have a lot of shortcuts and helpers to remove. 

> It's easier to explain and exposes less knobs to users. It can always be added later.

I think explaining that they have to go back and get another handle is more confusing than letting them set call parameters as parameters to the calls.

I think what we've done in TS is the most confusing approach of all.

How do you explain to a user how to set `first_execution_id` on cancel? Or on terminate?

How do you explain to a user how to query with one reject condition for one query and another reject condition for another?

Allowing parameters for a call to be set as parameters on the call is incredibly reasonable for these programming languages, we don't need to take that away and make them either go back and recreate the handle or tell them to go back and recreate the client or whatever else is just making things complicated.

Reply:
Against what I think is better judgement, the team has decided to remove call parameters from the parameters of the call :-(